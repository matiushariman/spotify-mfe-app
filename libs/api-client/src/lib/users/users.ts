import axios, { AxiosInstance, AxiosError, isAxiosError } from 'axios';

import { API_BASE_URL } from '../../constants';

import type {
  GetCurrentUserProfileResponse,
  GetCurrentUserProfileError,
} from './users.types';
import type { ApiClassConstructor } from '../../types/ApiClassConstructor.types';

export class UsersApi {
  private readonly http: AxiosInstance;

  constructor({ token }: ApiClassConstructor) {
    this.http = axios.create({
      baseURL: `${API_BASE_URL}/v1`,
      headers: {
        'Content-Type': 'application /json',
        Authorization: `Bearer ${token}`,
      },
    });
  }

  /**
   * Get detailed profile information about the current user (including the current user's username)
   *
   * - {@link https://developer.spotify.com/documentation/web-api/reference/#/operations/get-current-users-profile|API Reference - Get Current User Profile}
   */
  async getCurrentUserProfile() {
    return this.http
      .get<GetCurrentUserProfileResponse>('/me')
      .then((res) => res)
      .catch((err: Error | AxiosError<GetCurrentUserProfileError>) => {
        if (isAxiosError(err)) {
          return Promise.reject(err?.response?.data);
        }

        return Promise.reject(err);
      });
  }
}
