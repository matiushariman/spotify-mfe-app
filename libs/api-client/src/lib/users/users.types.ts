import type { ApiErrorResponse } from '../../types/ApiErrorResponse.types';
import type { ExternalUrls } from '../../types/ExternalUrls.types';
import type { Image } from '../../types/Image.types';
import type { Album } from '../../types/Album.types';
import type { Artist } from '../../types/Artist.types';
import type { ExternalIds } from '../../types/ExternalIds.types';
import type { Cursors } from '../../types/Cursors.types';
import type { Followers } from '../../types/Followers.types';

/**
 * - {@link https://developer.spotify.com/documentation/web-api/reference/#/operations/get-current-users-profile|API Reference - Get Current User Profile}
 */
export interface GetCurrentUserProfileResponse {
  /**
   * - The country of the users, as set in the users's account profile. An ISO 3166-1 alpha-2 country code.
   * - This field is only available when the current users has granted access to the users-read-private scope.
   */
  readonly country?: string;
  /**
   * - The name displayed on the users's profile. null if not available.
   */
  readonly display_name: string | null;
  /**
   * - The user's email address, as entered by the user when creating their account.
   * - This field is only available when the current users has granted access to the users-read-private scope.
   */
  readonly email?: string;
  /**
   * - The users's explicit content settings.
   * - This field is only available when the current users has granted access to the users-read-private scope.
   */
  readonly explicit_content?: {
    /**
     * - When true, indicates that explicit content should not be played.
     */
    filter_enabled: boolean;
    /**
     * - When true, indicates that the explicit content setting is locked and can't be changed by the users.
     */
    filter_locked: boolean;
  };
  /**
   * - Known external URLs for this users.
   */
  readonly external_urls: ExternalUrls;
  /**
   * - Information about the followers of the users.
   */
  readonly followers: Followers;
  /**
   * - A link to the Web API endpoint for this users.
   */
  readonly href: string;
  /**
   * - The Spotify users ID for the users.
   */
  readonly id: string;
  /**
   * - The users's profile image.
   */
  readonly images: Image[];
  /**
   * - The users's Spotify subscription level: "premium", "free", etc. (The subscription level "open" can be considered the same as "free".)
   * - This field is only available when the current users has granted access to the users-read-private scope.
   */
  readonly product: 'open' | 'free' | 'premium';
  /**
   * - The object type: "users"
   */
  readonly type: 'user';
  /**
   * - The Spotify URI for the users.
   */
  readonly uri: string;
}

export type GetCurrentUserProfileError = ApiErrorResponse;

export interface GetUserTopItemsRequest {
  /**
   * - The type of entity to return. Valid values: artists or tracks
   */
  readonly type: 'artists' | 'tracks';
  readonly query?: {
    /**
     * - The maximum number of items to return. Default: 20. Minimum: 1. Maximum: 50.
     */
    readonly limit: number;
  };
}

export interface UserTopTracks {
  readonly album: Album;
  readonly artists: Artist[];

  readonly available_markets: string[];
  readonly disc_number: number;
  readonly duration_ms: number;
  readonly explicit: boolean;
  readonly external_ids: ExternalIds;
  readonly external_urls: ExternalUrls;
  readonly href: string;
  readonly id: string;
  readonly is_local: boolean;
  readonly name: string;
  readonly popularity: number;
  readonly preview_url: string;
  readonly track_number: number;
  readonly type: 'track';
  readonly uri: string;
}

/**
 * - {@link https://developer.spotify.com/documentation/web-api/reference/#/operations/get-users-top-artists-and-tracks|API Reference - Get User's Top Items}
 */
export interface GetUserTopItemsResponse {
  readonly items: UserTopTracks[];
  /**
   * - A link to the Web API endpoint returning the full result of the request
   */
  readonly href: string;
  /**
   * - The maximum number of items in the response (as set in the query or by default).
   */
  readonly limit: number;
  /**
   * - URL to the next page of items. ( `null` if none)
   */
  readonly next: string | null;
  /**
   * - The offset of the items returned (as set in the query or by default)
   */
  readonly offset: number;
  /**
   * - URL to the previous page of items. ( `null` if none)
   */
  readonly previous: string | null;
  /**
   * - The total number of items available to return.
   */
  readonly total: number;
}

export type GetUserTopItemsError = ApiErrorResponse;

interface FollowedArtist {
  readonly name: string;
  readonly popularity: number;
  readonly type: 'artist';
  readonly uri: string;
  readonly href: string;
  readonly id: string;
  readonly external_urls: ExternalUrls;
  readonly followers: Followers;

  readonly genres: string[];

  readonly images: Image[];
}
export interface GetFollowedArtistsResponse {
  readonly artists: {
    /**
     * - A link to the Web API endpoint returning the full result of the request.
     */
    readonly href: string;
    /**
     * - The maximum number of items in the response (as set in the query or by default).
     */
    readonly limit: number;
    /**
     * - URL to the next page of items. ( `null` if none)
     */
    readonly next: string | null;

    readonly cursors: Cursors;
    /**
     * - The total number of items available to return.
     */
    readonly total: number;

    readonly items: FollowedArtist[];
  };
}

export type GetFollowedArtistsError = ApiErrorResponse;
