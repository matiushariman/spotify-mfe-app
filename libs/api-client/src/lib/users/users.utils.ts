import type { GetUserTopItemsRequest } from './users.types';
export const constructGetUserTopItemsQuery = (
  query?: GetUserTopItemsRequest['query']
) => {
  if (!query) {
    return '';
  }

  return `?limit=${query.limit}`;
};
