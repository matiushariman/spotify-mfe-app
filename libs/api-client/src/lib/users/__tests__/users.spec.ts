import { UsersApi } from '../users';

describe('lib/users', () => {
  const usersApi = new UsersApi({ token: '' });

  describe('getCurrentUserProfile()', () => {
    it('should return user profile data', async () => {
      const res = await usersApi.getCurrentUserProfile();

      expect(res.status).toEqual(200);
      expect(res.data).toMatchSnapshot();
    });
  });

  describe(usersApi.getUserTopItems, () => {
    it('should return user top tracks', async () => {
      const res = await usersApi.getUserTopItems({ type: 'tracks' });

      expect(res.status).toEqual(200);
      expect(res.data).toMatchSnapshot();
    });
  });
});
