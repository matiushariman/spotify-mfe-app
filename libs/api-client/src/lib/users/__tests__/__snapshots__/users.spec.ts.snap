// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lib/users getCurrentUserProfile() should return user profile data 1`] = `
Object {
  "country": "SG",
  "display_name": "Test Account",
  "email": "pottage.silt-0d@icloud.com",
  "explicit_content": Object {
    "filter_enabled": false,
    "filter_locked": false,
  },
  "external_urls": Object {
    "spotify": "https://open.spotify.com/user/31amcddtdt7u75pa3qzjukke6iiu",
  },
  "followers": Object {
    "href": null,
    "total": 0,
  },
  "href": "https://api.spotify.com/v1/users/31amcddtdt7u75pa3qzjukke6iiu",
  "id": "31amcddtdt7u75pa3qzjukke6iiu",
  "images": Array [],
  "product": "open",
  "type": "user",
  "uri": "spotify:user:31amcddtdt7u75pa3qzjukke6iiu",
}
`;

exports[`lib/users getUserTopItems should return user top tracks 1`] = `
Object {
  "href": "https://api.spotify.com/v1/me/shows?offset=0&limit=20
",
  "items": Array [
    Object {
      "album": Object {
        "album_type": "ALBUM",
        "artists": Array [
          Object {
            "external_urls": Object {
              "spotify": "https://open.spotify.com/artist/2iDVt6mFbtbDEZG5ax0dTi",
            },
            "href": "https://api.spotify.com/v1/artists/2iDVt6mFbtbDEZG5ax0dTi",
            "id": "2iDVt6mFbtbDEZG5ax0dTi",
            "name": "Tulus",
            "type": "artist",
            "uri": "spotify:artist:2iDVt6mFbtbDEZG5ax0dTi",
          },
        ],
        "available_markets": Array [
          "AD",
          "AE",
          "AR",
          "AT",
          "AU",
          "BE",
          "BG",
          "BH",
          "BO",
          "BR",
          "CA",
          "CH",
          "CL",
          "CO",
          "CR",
          "CY",
          "CZ",
          "DE",
          "DK",
          "DO",
          "DZ",
          "EC",
          "EE",
          "EG",
          "ES",
          "FI",
          "FR",
          "GB",
          "GR",
          "GT",
          "HK",
          "HN",
          "HU",
          "ID",
          "IE",
          "IL",
          "IN",
          "IS",
          "IT",
          "JO",
          "JP",
          "KW",
          "LB",
          "LI",
          "LT",
          "LU",
          "LV",
          "MA",
          "MC",
          "MT",
          "MX",
          "MY",
          "NI",
          "NL",
          "NO",
          "NZ",
          "OM",
          "PA",
          "PE",
          "PH",
          "PL",
          "PS",
          "PT",
          "PY",
          "QA",
          "RO",
          "SA",
          "SE",
          "SG",
          "SK",
          "SV",
          "TH",
          "TN",
          "TR",
          "TW",
          "US",
          "UY",
          "VN",
          "ZA",
        ],
        "external_urls": Object {
          "spotify": "https://open.spotify.com/album/3R4IAF9ApqYeUQrv1ddyoR",
        },
        "href": "https://api.spotify.com/v1/albums/3R4IAF9ApqYeUQrv1ddyoR",
        "id": "3R4IAF9ApqYeUQrv1ddyoR",
        "images": Array [
          Object {
            "height": 640,
            "url": "https://i.scdn.co/image/ab67616d0000b273b55d26c578e30129b0a7e86e",
            "width": 640,
          },
          Object {
            "height": 300,
            "url": "https://i.scdn.co/image/ab67616d00001e02b55d26c578e30129b0a7e86e",
            "width": 300,
          },
          Object {
            "height": 64,
            "url": "https://i.scdn.co/image/ab67616d00004851b55d26c578e30129b0a7e86e",
            "width": 64,
          },
        ],
        "name": "Manusia",
        "release_date": "2022-03-03",
        "release_date_precision": "day",
        "total_tracks": 10,
        "type": "album",
        "uri": "spotify:album:3R4IAF9ApqYeUQrv1ddyoR",
      },
      "artists": Array [
        Object {
          "external_urls": Object {
            "spotify": "https://open.spotify.com/artist/2iDVt6mFbtbDEZG5ax0dTi",
          },
          "href": "https://api.spotify.com/v1/artists/2iDVt6mFbtbDEZG5ax0dTi",
          "id": "2iDVt6mFbtbDEZG5ax0dTi",
          "name": "Tulus",
          "type": "artist",
          "uri": "spotify:artist:2iDVt6mFbtbDEZG5ax0dTi",
        },
      ],
      "available_markets": Array [
        "AD",
        "AE",
        "AR",
        "AT",
        "AU",
        "BE",
        "BG",
        "BH",
        "BO",
        "BR",
        "CA",
        "CH",
        "CL",
        "CO",
        "CR",
        "CY",
        "CZ",
        "DE",
        "DK",
        "DO",
        "DZ",
        "EC",
        "EE",
        "EG",
        "ES",
        "FI",
        "FR",
        "GB",
        "GR",
        "GT",
        "HK",
        "HN",
        "HU",
        "ID",
        "IE",
        "IL",
        "IN",
        "IS",
        "IT",
        "JO",
        "JP",
        "KW",
        "LB",
        "LI",
        "LT",
        "LU",
        "LV",
        "MA",
        "MC",
        "MT",
        "MX",
        "MY",
        "NI",
        "NL",
        "NO",
        "NZ",
        "OM",
        "PA",
        "PE",
        "PH",
        "PL",
        "PS",
        "PT",
        "PY",
        "QA",
        "RO",
        "SA",
        "SE",
        "SG",
        "SK",
        "SV",
        "TH",
        "TN",
        "TR",
        "TW",
        "US",
        "UY",
        "VN",
        "ZA",
      ],
      "disc_number": 1,
      "duration_ms": 242000,
      "explicit": false,
      "external_ids": Object {
        "isrc": "FR2X42253130",
      },
      "external_urls": Object {
        "spotify": "https://open.spotify.com/track/2hHeGD57S0BcopfVcmehdl",
      },
      "href": "https://api.spotify.com/v1/tracks/2hHeGD57S0BcopfVcmehdl",
      "id": "2hHeGD57S0BcopfVcmehdl",
      "is_local": false,
      "name": "Hati-Hati di Jalan",
      "popularity": 80,
      "preview_url": "https://p.scdn.co/mp3-preview/c5e70e742b9f7c131676f6cfb9c69e5fc4a4f85b?cid=06a2a95448cb464d870d7d98373601d9",
      "track_number": 8,
      "type": "track",
      "uri": "spotify:track:2hHeGD57S0BcopfVcmehdl",
    },
  ],
  "limit": 20,
  "next": "https://api.spotify.com/v1/me/shows?offset=1&limit=1",
  "offset": 0,
  "previous": "https://api.spotify.com/v1/me/shows?offset=1&limit=1",
  "total": 4,
}
`;
